runtime/commonTest/src/kotlinx/serialization/builtins/KeyValueSerializersTest.kt

```diff
@@ -0,0 +1,169 @@
+            """{"first":42,"second":null}"""
+        )
+    }
+
+    private inline fun <reified K, reified V> testPair(
+        pairInstance: Pair<K, V>,
+        kSerializer: KSerializer<K>,
+        vSerializer: KSerializer<V>,
+        useStreaming: Boolean,
+        expectedJson: String
+    ) {
+        run {
+            val serializer = PairSerializer(kSerializer, vSerializer)
+            val json = default.stringify(serializer, pairInstance, useStreaming)
+            assertEquals(expectedJson, json)
+            val pair = default.parse(serializer, json, useStreaming)
+            assertEquals(pairInstance, pair)
+        }
+
+        run {
+            val serializer = PairSerializer<K, V>()
+            val json = default.stringify(serializer, pairInstance, useStreaming)
+            assertEquals(expectedJson, json)
+            val pair = default.parse(serializer, json, useStreaming)
+            assertEquals(pairInstance, pair)
+        }
+    }
+
+    @Test
+    fun testTriple() = parametrizedTest { useStreaming ->
+        testTriple(
+            Triple(42, 42, "42"),
+            Int.serializer(),
+            Int.serializer(),
+            String.serializer(),
+            useStreaming,
+            """{"first":42,"second":42,"third":"42"}"""
+        )
+
+        testTriple(
+            Triple(42, Triple(42, "f", 'c'), "42"),
+            Int.serializer(),
+            TripleSerializer(),
+            String.serializer(),
+            useStreaming,
+            """{"first":42,"second":{"first":42,"second":"f","third":"c"},"third":"42"}"""
+        )
+
+        testTriple(
+            Triple(42, null, null),
+            Int.serializer(),
+            Int.serializer().nullable,
+            String.serializer().nullable,
+            useStreaming,
+            """{"first":42,"second":null,"third":null}"""
+        )
+    }
+
+    private inline fun <reified A, reified B, reified C> testTriple(
+        tripleInstance: Triple<A, B, C>,
+        aSerializer: KSerializer<A>,
+        bSerializer: KSerializer<B>,
+        cSerializer: KSerializer<C>,
+        useStreaming: Boolean,
+        expectedJson: String
+    ) {
+        run {
+            val serializer = TripleSerializer(aSerializer, bSerializer, cSerializer)
+            val json = default.stringify(serializer, tripleInstance, useStreaming)
+            assertEquals(expectedJson, json)
+            val triple = default.parse(serializer, json, useStreaming)
+            assertEquals(tripleInstance, triple)
+        }
+
+        run {
+            val serializer = TripleSerializer<A, B, C>()
+            val json = default.stringify(serializer, tripleInstance, useStreaming)
+            assertEquals(expectedJson, json)
+            val triple = default.parse(serializer, json, useStreaming)
+            assertEquals(tripleInstance, triple)
+        }
+    }
+
+    class Entry<K, V>(override val key: K, override val value: V) : Map.Entry<K, V> {
```

Comment:
`MapEntrySerializer` has its own `Map.Entry` impl. Probably you may reuse it. (also, that impl should be comparable with Map.Entry, so fix this please if you don't mind)

Reply:
It's private and I prefer to keep it that way.

>also, that impl should be comparable with Map.Entry

This is impossible because `equals` should be symmetric. Fun fact: `assertEquals(a, b)` works differently on different platforms (one invokes `a == b` and another one `b == a``)

Reply:
So you've overrided `equals` here only for testing purposes?

Reply:
yes